### PR TITLE
Anyhow removed in favour of log and thiserror

### LINKED ISSUE #93: Set default log level and remove `anyhow` in favour of `log` & `thiserror`

### PR DESCRIPTION
**Pull request by aadarsh-nagrath**
# Pull Request

## Changes Made
- Removed implementation and default/custom error variants of log, utilizing `log` and `thiserror` libraries.
- Implemented minor code improvements.
- Added dependencies for `log` and `thiserror`.
- Modified `bail`, `ensure`, and `context` methods of `anyhow` to use log implementation where necessary.

## Fixes
Fixes #93

## Additional Notes
/claim #93

Let me know if there are any concerns or further discussions needed. 👍

### CODE REVIEW

**Comment by koraa** [comment]
@wucke13 Can you review this please?

File: util/src/result.rs [hunk 1]
```diff
@@ -2,6 +2,11 @@
 #[macro_export]
 macro_rules! attempt {
     ($block:expr) => {
-        (|| -> ::anyhow::Result<_> { $block })()
+        (|| -> Result<_, ::rosenpass_util::SodiumError> {
```
Comment (koraa) [inline util/src/result.rs:5]: This is obviously not a good idea. Either infer the error type or get rid of the attempt macro

**Review by koraa — CHANGES_REQUESTED** [review]

**Comment by koraa** [comment]
@aadarsh-nagrath Thank you for your contribution!

File: rosenpass/benches/handshake.rs [hunk 2]
```diff
@@ -77,10 +93,12 @@ fn criterion_benchmark(c: &mut Criterion) {
     });
     c.bench_function("handshake", |bench| {
         bench.iter(|| {
-            hs(black_box(&mut a), black_box(&mut b)).unwrap();
+            if let Err(err) = hs(black_box(&mut a), black_box(&mut b)) {
+                error!("Handshake error: {:?}", err);
```
[the remark this answers is not in the document]
Comment (aadarsh-nagrath) [reply rosenpass/benches/handshake.rs:97]: I understand your concern about excessive logging and on benchmarks logs. My intention was to provide more visibility into the benchmark process, but I see the potential downsides you mentioned. 
I'm relatively new to open-source contributions, and I'm learning the best practices.

[the remark this answers is not in the document]
Comment (aadarsh-nagrath) [reply rosenpass/benches/handshake.rs:97]: I have added some changes

File: ciphers/src/subtle/incorrect_hmac_blake2b.rs [hunk 3]
```diff
@@ -18,27 +23,51 @@ pub const OUT_MAX: usize = blake2b::OUT_MAX;
...
 #[inline]
-pub fn hash<'a>(key: &'a [u8], data: &'a [u8]) -> impl To<[u8], anyhow::Result<()>> + 'a {
+pub fn hash<'a>(key: &'a [u8], data: &'a [u8]) -> impl To<[u8], Result<(), HmacError>> + 'a {
     const IPAD: [u8; KEY_LEN] = [0x36u8; KEY_LEN];
     const OPAD: [u8; KEY_LEN] = [0x5Cu8; KEY_LEN];
 
     with_destination(|out: &mut [u8]| {
+        if !log_enabled!(Level::Error) {
+            // Skip unnecessary computation if error logging is not enabled
+            return Err(HmacError::LoggingDisabled);
```
[the remark this answers is not in the document]
Comment (aadarsh-nagrath) [reply ciphers/src/subtle/incorrect_hmac_blake2b.rs]: logging check and the logging statement removed

File: util/src/result.rs (hunk 1 above)
Comment (aadarsh-nagrath) [reply util/src/result.rs]: infered the error type

**Comment by koraa** [comment]
@aadarsh-nagrath Thank you for continuing to work on this! I will check with @wucke13 regarding his view on this…

**Comment by wucke13** [comment]
@aadarsh-nagrath Please rebase with `main` and fix the merge conflicts. If any checks in CI fail, fix them. Then we can merge, and you get the bounty. I suggest that we increase the bounty to 50$ as well. @koraa, are you ok with that?

**Comment by koraa** [comment]
> /bounty 20

Doesn't work. @aadarsh-nagrath please do remind me to leave a $20 tip…

**Comment by koraa** [comment]
@tushar-haldia Would you prefer I close this PR?

**Comment by prabhpreet** [comment]
@aadarsh-nagrath Checking in on this PR

> @aadarsh-nagrath Please rebase with `main` and fix the merge conflicts. If any checks in CI fail, fix them. Then we can merge, and you get the bounty. I suggest that we increase the bounty to 50$ as well. @koraa, are you ok with that?

**Comment by prabhpreet** [comment]
Closing this PR due to inactivity.

**Comment by aadarsh-nagrath** [comment]
> Sure, sounds good! And thank you for the good work @aadarsh-nagrath!

👍

**Pull request closed without merging**